### DISCUSSION TITLE
Bad request while sending document

### DISCUSSION THREAD

**Discussion by vinoddanekar** [OxalisCommunity/oxalis#609] [Q&A]
Hi all,

I have a weird scenario.
Receiving docs on same server - works
Receiving docs from remote participants - works
Sending docs to participants present on same server - works
**Sending docs to remote participants - fails with bad request.**

When we did peppol test bench test then we could send docs on it. Currently, I was testing it with NPA, it did not work

version.oxalis: 5.0.6
version.java: 1.8.0_241
mode: TEST

Logs are as below:
![image](https://user-images.githubusercontent.com/46052955/186909926-607b102a-c1eb-4d19-b745-6b3670cce545.png)
![image](https://user-images.githubusercontent.com/46052955/186910098-a6ecaccc-5185-4296-868e-06cd1c65b007.png)

Please, guide me through. Unable to understand what could be wrong.

Thanks

Vinod

**Comment by dladlk** [comment]
Try to activate CXF debug logging to see which XML is generated and sent to your test AP https://test.peppolautoriteit.nl/as4_test (looks like [Ionite.net](https://ionite.net/newsblog/2022-02-04_ion_ap_ion_smp_cef_conformant/) Access Point). You can do it by setting logging category 

org.apache.cxf.services.DummyImpl

to INFO - or even the whole org.apache.cxf

**Comment by aaron-kumar** [comment]
@vinoddanekar  : Please make sure that request is created correctly. You can use detailed debugging as e.g. suggested by @dladlk .

As for as Oxalis and NPA (https://test.peppolautoriteit.nl/as4_test) is concerned then everything is fine. Messages are being delivered successfully. I used same endpoint as mentioned in NPa Peppol Test Tool for performing test with  Oxalis/Oxalis 5.4.0

Request OUT
.....
 o.a.cxf.services.DummyImpl.REQ_OUT - REQ_OUT
    Address: https://test.peppolautoriteit.nl/as4_test
    HttpMethod: POST
    Content-Type: multipart/related; type="application/soap+xml"; boundary="uuid:a74aa0fd-89fd-45d6-be9c-03fa7bf2ea5f"; start="<root.message@cxf.apache.org>"; start-info="application/soap+xml"
    ExchangeId: 9387520b-15c5-48d7-bbab-95cfc796ead2
    ServiceName: outbound-service
    PortName: port
    PortTypeName: DummyImpl
    Headers: {Accept=*/*, User-Agent=Oxalis 5.4.0 / AS4 5.4.0 / 
......

Response IN:
.....
 o.a.cxf.services.DummyImpl.RESP_IN - RESP_IN
    Address: https://test.peppolautoriteit.nl/as4_test
    Content-Type: text/xml
    ResponseCode: 200
    ExchangeId: 9387520b-15c5-48d7-bbab-95cfc796ead2
    ServiceName: outbound-service
    PortName: port
    PortTypeName: DummyImpl 
......

**Reply by vinoddanekar** [reply]
Thanks for reply. I will enable debug as suggested by @dladlk and will respond you here.

Do you feel that upgrading to 5.4 might solve the issue? Is so then I will opt for it. I am not much familiar with Java so have little struggles in finding and resolving the errors.

Thanks in advance.

**Comment by aaron-kumar** [comment]
@vinoddanekar  : All Oxalis release builds properly tested, so it is Not Oxalis version specific thing. 
In general, we recommend to use latest version, as latest version contains functionality enhancement and lot of security vulnerabilities fix. 

Note: when upgrading make sure that you use compatible version of Oxalis and Oxalis-AS4  (Latest is 5.4.0 for both)

**Reply by vinoddanekar** [reply]
@aaron-kumar 

You correctly said that its no version specific. Still, I upgraded to 5.4.0 both Oxalis and AS4.

Following 3 errors I find (in earlier version too)
- Says one reference is missing
![image](https://user-images.githubusercontent.com/46052955/188124318-b7569b71-763d-4828-8e83-58ec5d800749.png)

- Its resolving something with guid@server-name but this value was not allocated by me. seems autogenerated
![image](https://user-images.githubusercontent.com/46052955/188124384-15b99e5f-5333-44b3-b353-ecaf26775f4a.png)

- Response throws HTTP 500
![image](https://user-images.githubusercontent.com/46052955/188124419-52e201be-2244-4e3b-bcb1-912076fa0d6a.png)

REQ_OUT
![image](https://user-images.githubusercontent.com/46052955/188125416-e452b0b5-e3f7-48f6-bd43-086da3edfa7d.png)

RESP_IN
![image](https://user-images.githubusercontent.com/46052955/188125756-df813a24-f5be-415d-a909-2a685f545f1d.png)

Oxalis release used: https://github.com/OxalisCommunity/oxalis/releases/tag/v5.4.0
AS4 release used: https://github.com/OxalisCommunity/Oxalis-AS4/releases/tag/v5.4.0

Unable to trace the missing reference and something its trying to resolve guid@server-name. 

Can you please help me resolving these things?

**Reply by dladlk** [reply]
In your REQ_OUT there should be an actual payload sent - XML itself. Try to look there - it should be something wrong with this XML. You can post it here is it does not contain private info. Your screenshot for REQ_OUT ends just before the payload itself :)

It should be something like this:

![image](https://user-images.githubusercontent.com/41484081/188152250-4b2924b4-028f-40a6-bb95-e230b2ff5ff7.png)

Try to save the whole line as XML and check the document structure, it should help.

**Reply by vinoddanekar** [reply]
Hi @dladlk, I can see the Envelope XML and checked opening it. XML structure looks OK to me. 

Can you please let me know if these two errors are culprit and how to solve them:
1. cant load class com.sun.xmlmessaging.saaj.soap.SOAPDocumentImpl
2. I was asked to resolve ...., returns null
  present in screenshot 1, 2.

**Reply by dladlk** [reply]
Hello @vinoddanekar  - I have never seen such exceptions in our envs, so it is hard to say what exactly provoked them.
But of cause it can relate to the issue you experience, although I hardly believe anyone can tell you what exactly is wrong in your setup :)

SOAPDocumentImpl is in saaj-impl-1.4.0-b03.jar (at least in Oxalis 5.4.0, you use much older release) - you can check that you have it in your classpath. But it had to be used also for local sending - which you said works ok, so it is weird why it works locally and fails on remote sending.

SAAJ stays for **S**OAP with **A**ttachments **A**PI for **J**ava - and in AS4 the encrypted payload is sent as attachment to SOAP Envelope, so after the line with env:Envelope there should be some encrypted binary attachment - and as you get this error with absent class, it can be the reason why it fails on sending the message.

I guess these exceptions relate to each other - so I would recommend to start with finding out why SOAPDocumentImpl is not found in your classpath.

**Reply by vinoddanekar** [reply]
Hi @dladlk - Thanks for your insights. It helped me removing the saaj error. I see that in latest release 5.4 https://github.com/OxalisCommunity/oxalis/releases/tag/v5.4.0 this jar file is not present. I tried to retrieve it from older version from backsups. I got 1.3 and it removed the error. So, at least a little progress. :)

Second error where it says "I was asked whether I can resolve guid@server-name". Can you please tell me what its really trying to resolve so that I can check respective area - may be some config?

![image](https://user-images.githubusercontent.com/46052955/188438974-5274688e-a93d-4965-b290-08dcd2a0abb1.png)

Thanks for the help.

**Reply by dladlk** [reply]
Oh... Wrong fix - but helped to understand what is wrong - I guess :)

SOAPDocumentImpl is taken from saaj-impl-1.4.0-b03.jar only in single case - when CXF runs under non Java-8 :)

Because com.sun.xml.internal.messaging.saaj.soap.SOAPDocumentImpl is a part of Java 8 standard libraries - and e.g. in Oracle/Sun JDK it is in rt.jar :)

**Almost 100% probability you run not Java 8 actually :)))**

Please remove this jar and ensure that the java version is 8 - as in later versions a lot of javax./com.sum are removed from JDK, and we are still in progress to support them in similar way as CFX does it - by embedding jars which contain these classes if it is non Java8 environment.

**Reply by dladlk** [reply]
For history - CXF has a profile Java9-plus which is activated starting from Java 9 - and adds dependency on saaj-impl among many other removed dependencies:

https://github.com/apache/cxf/blob/8d5c0320bedd710fcba82092acdcca35b25e7ac0/parent/pom.xml#L2827:L2892

**Reply by vinoddanekar** [reply]
Java version screenshot:
![image](https://user-images.githubusercontent.com/46052955/188462679-d858e6aa-2760-4ffb-b972-1c2212e825b7.png)

From Oxalis web page:
version.oxalis: 5.4.0
version.java: 1.8.0_241
mode: TEST
certificate.expired: false
build.id: e987e4a76251bc283e0d8e1b8a2d5bbbe7eab4c5
build.tstamp: 02.08.2022 @ 15:00:14 UTC

I did more digging -
- When I send on same server with certificate - works fine (interesting fact is that I get NULL for "I was asked whether I can resolve guid@server-name" still it works)
- When I send on same server without certificate - fails saying no transmission provided
- When I send to NPA without certificate - HTTP 500
- When I send to NPA with certificate - HTTP 400

Im not 100% sure but I dont think its a java version problem. Is there any other free test tool I can use to ensure if its prominent problem?

I appreciate your help.

**Reply by dladlk** [reply]
Are you sending and receiving from the same application? Usually, it is done in separate. I would suppose, that you make a screen from Oxalis web page of server (receiving) - but send from another application. 
What is your integration of oxalis? How do you run the sending process? If you use oxalis-standalone sub-distribution, then it can have another java then the server has: https://github.com/OxalisCommunity/oxalis/blob/a037409cf0b06e582786e0a182220d099176460b/oxalis-dist/oxalis-standalone/example.bat#L8

But if you integrate oxalis as4 as dependency into your own application, just add logging of java.version system property to be absolutely sure - like ```System.out.println(System.getProperty("java.version"));``` 

Again, I think you run >Java 8 because of ClassNotFoundException for SOAPDocumentImpl - which is a part of rt.jar. You can check jars in your Java installation for this file SOAPDocumentImpl.class - what if you have "corrupted" installation :) ?

**Reply by vinoddanekar** [reply]
For AS2 earlier, we were using jar directly but when AS4 introduced we are using it like this -

` java -classpath "standalone/*;AS4/*" eu.sendregning.oxalis.Main -f Test_NPA.xml`

Here in folder standalone - I have standlone.jar and in folder AS4 - I gave AS4 plugin.

Internally while sending I make use of parameters: -cert for certificate and -protocol for protocol and -u for AP url.

**Reply by dladlk** [reply]
Try to replace java with a direct path to java 8 - e.g. "C:\Program Files\Java\jdk1.8.0_241\bin\java.exe"...

**Reply by vinoddanekar** [reply]
Java is pointed to the same. also tried explicitly as you mentioned - no difference.

**Reply by dladlk** [reply]
Oh... It becomes very interesting : ) 
Try to check if you have set JAVA_HOME to something from java12... Do you have such env variable on your computer in the context of your console?
```echo %JAVA_HOME%```

**Reply by dladlk** [reply]
As I understand, you use Oracle JDK. Try to check the installation folder rt.jar - it should contain the class which Oxalis cannot load: \com\sun\xml\internal\messaging\saaj\soap\SOAPDocumentImpl.class in last free version of Oracle Java 8 "C:\Program Files\Java\jdk1.8.0_202\jre\lib\rt.jar"

As I understand, you just unpacked standalone and AS4 folders to your current folder. 

Can you make ```dir /S``` to list files with sizes in your work folder (just delete from listing too specific files)?
